network/src/peer_manager/mod.rs

```diff
@@ -348,51 +348,59 @@ where
     fn handle_connection_event(&mut self, event: ConnectionNotification<TSocket>) {
         trace!("ConnectionNotification::{:?}", event);
         match event {
-            ConnectionNotification::NewConnection(identity, addr, origin, conn) => {
+            ConnectionNotification::NewConnection(conn) => {
+                info!("New connection established: {:?}", conn,);
+                // Update libra_network_peer counter.
+                counters::LIBRA_NETWORK_PEERS
+                    .with_label_values(&[self.role.as_str(), "connected"])
+                    .inc();
+                self.add_peer(conn);
+            }
+            ConnectionNotification::Disconnected(lost_conn_metadata, reason) => {
+                // See: https://github.com/libra/libra/issues/3128#issuecomment-605351504 for
+                // detailed reasoning on `Disconnected` events should be handled correctly.
                 info!(
-                    "{:?} connection with peer {} established",
-                    origin,
-                    identity.peer_id().short_str()
+                    "Connection {:?} closed due to {:?}",
+                    lost_conn_metadata, reason,
                 );
-                self.add_peer(identity, addr, origin, conn);
-            }
-            ConnectionNotification::Disconnected(identity, addr, origin, reason) => {
-                // If we receive a PeerDisconnected event with reason
-                // DisconnectReason::Requested or of a different origin from that of the
-                // currently active connection, this particular event is from a Peer actor
-                // which is being shutdown due to simultaneous dial tie-breaking and we don't
-                // need to send a LostPeer notification to all subscribers.
-                match reason {
-                    DisconnectReason::ConnectionLost => {
+                // Update libra_network_peer counter.
+                counters::LIBRA_NETWORK_PEERS
+                    .with_label_values(&[self.role.as_str(), "connected"])
+                    .dec();
```

Comment:
why would we decrement if we don't successfully remove a peer from active_peers? why is this not in the if statement on line 378?

Reply:
Here's the problem - if the disconnect was initiated by the client, the connection_id will not be in the `active_peers` map. It would have been moved to `outstanding_disconnect_requests` when the PeerManager received the request. So we go with the approach of incrementing connected_peers count on each successful connection and decrementing it on each disconnect.